plugins/auto-sizes/hooks.php

```diff
@@ -78,6 +78,26 @@ function auto_sizes_update_content_img_tag( $html ): string {
 }
 add_filter( 'wp_content_img_tag', 'auto_sizes_update_content_img_tag' );
```

Comment:
While from a DX perspective I agree, there are potential performance implications for it, given how often this is invoked. I believe some explorations in that regard were made in the past that surfaced some problems, but I don't recall the exact specifics.

Either way, I think that the right place to use `WP_HTML_Tag_Processor` for this is in WP core itself, and then introduce new filters that deal with the specific tags and attributes parsed by the processor. I don't think instantiating a new instance of the class for every single image in the content is a good approach.

Reply:
It does have better DX, but it's not really about that. It's about robustness, safety, and ensuring the result is correct. Instantiating the `WP_HTML_Tag_Processor` is very lightweight. It's being used increasingly throughout core as this is its intended purpose. We're also using it in our Performance plugins including this one:

https://github.com/WordPress/performance/blob/e7313851f39749886a4ae0bacbcd620a6effce3d/plugins/auto-sizes/hooks.php#L122-L141

https://github.com/WordPress/performance/blob/e7313851f39749886a4ae0bacbcd620a6effce3d/plugins/auto-sizes/hooks.php#L153-L157

https://github.com/WordPress/performance/blob/e7313851f39749886a4ae0bacbcd620a6effce3d/plugins/webp-uploads/picture-element.php#L84-L89

https://github.com/WordPress/performance/blob/e7313851f39749886a4ae0bacbcd620a6effce3d/plugins/webp-uploads/hooks.php#L542-L562

If there are problems with HTML Tag Processor, we should surface them to fix in core. Otherwise, we're using it elsewhere for this purpose, so I think we should use it here too.

Reply:
I'm basically channeling @dmsnell right now 😂

Reply:
Extra performance overhead was true of the first iteration of the HTML Tag Processor, in WP 6.4 (I believe). However, a lot has been improved since then. I don't think we've conducted a thorough analysis of whether using it in cases like this still has potential negative performance implications. Given how useful this is and how quickly I've seen adoption throughout Core, I would be in support of us doing some additional analysis. In the mean time, I think we should use it where we can so we're dogfooding our own WP Core APIs.

Reply:
> Either way, it's a separate change that should be separately discussed. :)

IMO, it's not really a separate change because it is also about hardening the logic add the `auto` keyword to the `sizes` attribute. Here's a test case where duplication occurs:

```html
<img
    data-tshirt-sizes="S M L"
    src="https://pd.w.org/2024/08/36566abfcb2952e20.99062056-300x225.jpg"
    srcset="https://pd.w.org/2024/08/36566abfcb2952e20.99062056-300x225.jpg 300w, https://pd.w.org/2024/08/36566abfcb2952e20.99062056-1024x768.jpg 1024w, https://pd.w.org/2024/08/36566abfcb2952e20.99062056-768x576.jpg 768w, https://pd.w.org/2024/08/36566abfcb2952e20.99062056-1536x1152.jpg 1536w, https://pd.w.org/2024/08/36566abfcb2952e20.99062056-2048x1536.jpg 2048w"
    sizes="auto, (max-width: 650px) 100vw, 650px"
    loading="lazy"
>
```

This results in:

```html
<img decoding="async"
    data-tshirt-sizes="auto, S M L"
    src="https://pd.w.org/2024/08/36566abfcb2952e20.99062056-300x225.jpg"
    srcset="[https://pd.w.org/2024/08/36566abfcb2952e20.99062056-300x225.jpg 300w](https://pd.w.org/2024/08/36566abfcb2952e20.99062056-300x225.jpg),[ https://pd.w.org/2024/08/36566abfcb2952e20.99062056-1024x768.jpg 1024w](https://pd.w.org/2024/08/36566abfcb2952e20.99062056-1024x768.jpg),[ https://pd.w.org/2024/08/36566abfcb2952e20.99062056-768x576.jpg 768w](https://pd.w.org/2024/08/36566abfcb2952e20.99062056-768x576.jpg),[ https://pd.w.org/2024/08/36566abfcb2952e20.99062056-1536x1152.jpg 1536w](https://pd.w.org/2024/08/36566abfcb2952e20.99062056-1536x1152.jpg),[ https://pd.w.org/2024/08/36566abfcb2952e20.99062056-2048x1536.jpg 2048w](https://pd.w.org/2024/08/36566abfcb2952e20.99062056-2048x1536.jpg)"
    sizes="auto, auto, (max-width: 650px) 100vw, 650px"
    loading="lazy"
>
```

In any case, I've opened #1446 to continue this, as well as to document other test cases which fail with the current implementation of `auto_sizes_update_content_img_tag()`.